NthKey/Model/SubscriptionManager.swift

```diff
@@ -7,13 +7,240 @@
 //  license, see the accompanying file LICENSE.md
 
 import Foundation
+import Combine
+import StoreKit
 
-final class SubscriptionManager: ObservableObject {
-    @Published var hasSubscription: Bool = false
+final class SubscriptionManager: NSObject, ObservableObject {
+    @Published private(set) var hasSubscription: Bool = false
+    @Published private(set) var products: [SKProduct] = []
+
+    // Used for direct user notifications about purchase flow
+    let purchasePublisher = PassthroughSubject<(String, Bool), Never>()
+
+    static var canMakePayments: Bool {
+        SKPaymentQueue.canMakePayments()
+    }
+
+    private var totalRestoredPurchases = 0
+
+    private let identifiers: [String]
+    private let sharedSecret: String?
+
+    init(identifiers: [String], sharedSecret: String? = nil) {
+        self.identifiers = identifiers
+        self.sharedSecret = sharedSecret
+        super.init()
+
+        checkPurchaseStatus()
+    }
+
+    // MARK: - Public
+
+    func startObserving() {
+        SKPaymentQueue.default().add(self)
+    }
+
+    func stopObserving() {
+        SKPaymentQueue.default().remove(self)
+    }
+
+    func prepareData() {
+        getProducts()
+    }
+
+    func purchase(product: SKProduct) -> Bool {
+        guard SubscriptionManager.canMakePayments else { return false }
+
+        let payment = SKPayment(product: product)
+        SKPaymentQueue.default().add(payment)
+
+        return true
+    }
+
+    func restorePurchases() {
+        totalRestoredPurchases = 0
+        SKPaymentQueue.default().restoreCompletedTransactions()
+    }
+
+    // MARK: - Private
+
+    private func getProducts() {
+        let request = SKProductsRequest(productIdentifiers: Set(identifiers))
+        request.delegate = self
+        request.start()
+    }
+
+    private func checkPurchaseStatus() {
+        guard let date = UserDefaults.subscriptionDate else { return }
+        hasSubscription = Date() < date
+    }
+
+    /// Bad practice apple.com/documentation/storekit/in-app_purchase/validating_receipts_with_the_app_store
+    /// Do not call the App Store server verifyReceipt endpoint from your app...
```

Comment:
Can you make an issue for this? We can fix it after the next TestFlight release, but maybe before the App Store release.

Reply:
added https://github.com/Sjors/nthkey-ios/issues/46